router/fastdp.go

```diff
@@ -699,19 +703,33 @@ func (fwd *fastDatapathForwarder) ErrorChannel() <-chan error {
 	return fwd.errorChan
 }
 
+func (fwd *fastDatapathForwarder) HealthChannel() <-chan bool {
+	return fwd.healthChan
+}
+
 func (fwd *fastDatapathForwarder) doHeartbeats() {
 	var err error
-
 	for err == nil {
 		select {
 		case <-fwd.heartbeatTimer.C:
 			if fwd.confirmed {
+				log.Debug(fwd.logPrefix(), "sending Heartbeat to peer")
 				fwd.sendHeartbeat()
 			}
 			fwd.heartbeatTimer.Reset(fwd.heartbeatInterval)
 
 		case <-fwd.heartbeatTimeout.C:
-			err = fmt.Errorf("timed out waiting for vxlan heartbeat")
+			log.Debug(fwd.logPrefix(), "missed Heartbeat from peer, marking fastdp forwarder as un-healthy")
+
+			// treat missed heartbeat as transient error. Indicate to overlay forwarder
+			// that Forwarder is un-healthy so it can pick next best forwarder
+			fwd.healthChan <- false
+
+			if fwd.healthy {
+				fwd.healthy = false
+			} else {
+				err = fmt.Errorf("timed out waiting for vxlan heartbeat")
```

Comment:
I can't see how we will ever hit this line. `healthy` starts off `true` and is reset to `true` when we reset the timer.
I can see that you don't want an error which will shut down the forwarder, but you should have some replacement log line.

And, if `err` is never set, then we don't need the surrounding logic to handle it.

Reply:
I should have expressed the intent more clearly. What `else` block does is if a forwarder is not healthy already and we have a consecutive heartbeat timeout then that forwarder is shutdown permanently. Which is not really better approach (just waiting till 2 heartbeatTimeout before shutting down the forwarder)

So if a forwarder is not `healthy` then should we have a time bound till when weaver keep sending the hearbeats? or should we keep sending the heartbeats forever without shutting down the forwarder?

Reply:
I made the change so that forwarder keep trying heartbeats and never shutdowns.